src/curl.cpp

```diff
@@ -703,7 +703,12 @@ size_t S3fsCurl::DownloadWriteCallback(void* ptr, size_t size, size_t nmemb, voi
     }
     pCurl->partdata.startpos += totalwrite;
     pCurl->partdata.size     -= totalwrite;
-
+    
+    // flush the file and clean the page cache
+    if (pCurl->partdata.size == 0){
+        fdatasync(pCurl->partdata.fd);
```

Comment:
Why is this desirable?  The kernel should flush its cache as needed and due to memory pressure.  Maybe you should configure your system differently if this is the behavior your want.

Reply:
Is this a fix for #2156?

Reply:
yes,  this fix make the s3fs use less  system page cache to write the big file to download， in my test， if do not add fdatasync， the cache use  will be 8GB-11GB when  copy a 13GB file from the dir of s3fs mount to  local disk。

Reply:
@tanguofu 
(I may be slightly wrong)
When an object is downloaded with s3fs, the size of the temporary file used for the CURL buffer is small because it should be downloaded in pieces instead of downloading the entire object size at once.
So I don't think this fix will make a big difference in memory usage.
(But, it recycles CURL handles, so that may be affecting it.)
However, other issues also point out memory usage, so I can't deny the cause of this correction.
Also, what are your s3fs boot options? (I want to confirm if it is related)

Reply:
i start s3fs use with options: `-o tmpdir=/tmpdir -o use_cache=/cachedir`; the memory is used by  `active_file` cache, this is crop from the oom log of the pod:
![oom log](https://github.com/s3fs-fuse/s3fs-fuse/assets/87045830/3d4233f2-d614-4e42-b796-25a40bc34196)

Reply:
@tanguofu Thanks for your reply.
Let me ask you some questions.
I assume you are using s3fs in a docker (or kubernetes) container (pod) by #2156.
Is the partition specified for this `tmpdir` and `use_cache` a disk using `tmpfs`?

What worries me is the following:
The s3fs cache file will continue to be created unless you specify a space limit for it, so `tmpfs`=`memory` will keep increasing.

Would calling `fdatasync` like in this PR code reduce the memory size used?
(I think the dirty disk cache can be reduced, but the s3fs cache file remains in the cache diretory(`tmpfs`), so I think the memory size will eventually overflow.)

Could you clear the OOM by specifying the upper limit of the usage capacity of this cache file, such as by using the `ensure_diskfree` option, or by not using the cache file(or deleting it periodically by an external process)?

Reply:
the `tmpdir` and `use_cache`  just store at the disk of the pod dir  instead of `tmpfs`;  which can be very by `df`;
 
i think this is easy to reproduction if  copy the large file in the pod  from mount s3fs dir to local dir; i test in the centos 7.9 os and k8s 1.18 version;

by call `fdatasync` in the download write function of curl write callback , the memory use of the pod is reduced from 10GB to the 10MB approximately, this is  verified   by `kubectl top pod -n xxx  xxx`；

our project we just use this patch and now the memory use is very little right now，i think use `fdatasync` to clear cache  take less  negative effects； at my test this make little effect on performance.

as far as know another choice is drop system cache by `sysctl -w vm.drop_caches=1` but this has more effect on performance.

Reply:
@tanguofu Thanks for the explanation, I'm thinking of merging this.
(Because there are issue reports that the amount of memory used increases, in addition to your case)

However, after the merge I will add an option(ex. `need_datasync`).
It will make this process will only be enabled when explicitly specified that option, rather than the default behavior.

Reply:
I strongly disagree with this unnecessary behavior.  Which other tools do this kind of proactive syncing?  I think it is misleading to users even if added as an option.  Suggest reverting before 1.93.

Reply:
If you want to hack your programs to unnecessarily sync you should do this via an `LD_PRELOAD` hack and not pollute random programs.